### ISSUE TITLE
Conflicting errors reported in Squiz standard for spaces before double arrow

### ISSUE THREAD

**Issue by nkovacs** [squizlabs/PHP_CodeSniffer#1033]
The following file:

```php
<?php

$arr = [
        'foo'=> 'bar',
        'bazbaz' => 'qux',
       ];
```

Generates two conflicting errors using the Squiz standard:

- Array double arrow not aligned correctly; expected 4 space(s) but found 0 (Squiz.Arrays.ArrayDeclaration.DoubleArrowNotAligned)
- Expected 1 space before "=>"; 0 found (Squiz.WhiteSpace.OperatorSpacing.NoSpaceBefore)

using phpcs 2.6.1

**Comment by aik099** [comment]
Will it be better if you'd use regular array syntax?

**Comment by nkovacs** [comment]
No, it's the same.

**Comment by aik099** [comment]
But why you're saying errors are conflicting?

* `Squiz.Arrays.ArrayDeclaration.DoubleArrowNotAligned` means that all `=>` must be in same column (no matter if spaces are needed before/after them or not)
* `Squiz.WhiteSpace.OperatorSpacing.NoSpaceBefore` says that space is needed before `=>`

**Comment by nkovacs** [comment]
Because `Squiz.WhiteSpace.OperatorSpacing.NoSpaceBefore` says that there should be 1 space before `=>`, and I understood that to mean exactly one space (the sniff actually requires at least one space). It's also redundant.

**Comment by gsherwood** [comment]
It's not that the standard conflicts. It's just that the array sniff supersedes the more generic whitespace sniff in this case.

Sniffs are designed to be generic and reused in other standards so they very rarely make assumptions about what other sniffs might be running.

Your also find that sometimes you fix one error and another errors pops up because a sniff is only doing as much work as it needs to give you that first message. But if you tell PHPCS to report a diff, or ask PHPCBF to auto-fix the file, you'll see that it sorts itself out. That's why both tools loop over the file more than once while diffing and fixing - to fix additional issues that the original fixes caused.

In this very specific case, the diff looks like this:
```diff
$ phpcs temp.php --standard=Squiz --report=diff
--- temp.php
+++ PHP_CodeSniffer
@@ -1,5 +1,5 @@
 <?php
 $arr = [
-        'foo'=> 'bar',
+        'foo'    => 'bar',
         'bazbaz' => 'qux',
        ];
```
You can see that the end result shows the double arrow in the correct position.

If you want to see how this specific conflict is resolved, the raw output is:
```
---START FILE CONTENT---
1|<?php
2|$arr = [
3|        'foo'=> 'bar',
4|        'bazbaz' => 'qux',
5|       ];
6|
--- END FILE CONTENT ---
	E: [Line 3] Array double arrow not aligned correctly; expected 4 space(s) but found 0 (Squiz.Arrays.ArrayDeclaration.DoubleArrowNotAligned)
	Squiz_Sniffs_Arrays_ArrayDeclarationSniff (line 768) replaced token 8 (T_CONSTANT_ENCAPSED_STRING) "'foo'" => "'foo'····"
	E: [Line 3] Expected 1 space before "=>"; 0 found (Squiz.WhiteSpace.OperatorSpacing.NoSpaceBefore)
	Squiz_Sniffs_WhiteSpace_OperatorSpacingSniff (line 245) replaced token 9 (T_DOUBLE_ARROW) "=>" => "·=>"
	* fixed 2 violations, starting loop 2 *
---START FILE CONTENT---
1|<?php
2|$arr = [
3|        'foo'     => 'bar',
4|        'bazbaz' => 'qux',
5|       ];
6|
--- END FILE CONTENT ---
	E: [Line 3] Array double arrow not aligned correctly; expected 4 space(s) but found 5 (Squiz.Arrays.ArrayDeclaration.DoubleArrowNotAligned)
	Squiz_Sniffs_Arrays_ArrayDeclarationSniff (line 770) replaced token 9 (T_WHITESPACE) "·····=>" => "····=>"
	* fixed 1 violations, starting loop 3 *
---START FILE CONTENT---
1|<?php
2|$arr = [
3|        'foo'    => 'bar',
4|        'bazbaz' => 'qux',
5|       ];
6|
--- END FILE CONTENT ---
```
If you read that, you'll see that both fixes were applied on the first loop, but we then end up with 1 extra space, which is corrected on the second loop. That's how sniffs that seemingly conflict sort themselves out.

The other thing to consider with this particular whitespace sniff is that it is trying to enforce a single space before the operator. But it does make an assumption about assignments allowing multiple spaces (like within arrays or just alignment assignment blocks) and so it doesn't check the number of spaces before them. But if it sees that there is no space, it knows that is definitely an error and so it reports and fixes it.

It needs to be this way to ensure sniffs are independently useful in other standards. So this isn't something I'm going to change in the Squiz standard, but hopefully I've explained why both errors are reported here.
